sys/usb/usbus/usbus_control.c

```diff
@@ -0,0 +1,417 @@
+                if (usbus_control_slicer_nextslice(usbus)) {
+                    _recv_setup(usbus, ep0_handler);
+                    ep0_handler->setup_state = USBUS_SETUPRQ_INDATA;
+                }
+                else {
+                    /* Ready out ZLP */
+                    usbdev_ep_ready(ep0_handler->out, 0);
+                    ep0_handler->setup_state = USBUS_SETUPRQ_OUTACK;
+                }
+            }
+            break;
+        case USBUS_SETUPRQ_OUTDATA:
+            if (ep->dir == USB_EP_DIR_OUT) {
+                /* Ready in ZLP */
+                ep0_handler->setup_state = USBUS_SETUPRQ_INACK;
+                size_t len = 0;
+                usbdev_ep_get(ep, USBOPT_EP_AVAILABLE, &len, sizeof(size_t));
+                DEBUG("Expected len: %d, received: %d\n",
+                      ep0_handler->setup.length, len);
+                if (ep0_handler->setup.length == len) {
+                    DEBUG("DATA complete\n");
+                    usbdev_ep_ready(ep0_handler->in, 0);
+                }
+                /* Flush OUT buffer */
+                usbdev_ep_ready(ep0_handler->out, 0);
+            }
+            else {
+                DEBUG("usbus_control: Invalid state OUTDATA with IN request\n");
+            }
+            break;
+        case USBUS_SETUPRQ_READY:
+            if (ep->dir == USB_EP_DIR_OUT) {
+                memset(&ep0_handler->slicer, 0, sizeof(usbus_control_slicer_t));
+                memcpy(&ep0_handler->setup, ep0_handler->out->buf,
+                       sizeof(usb_setup_t));
+                ep0_handler->slicer.reqlen = ep0_handler->setup.length;
+                _recv_setup(usbus, ep0_handler);
+            }
+            else {
+                DEBUG("usbus_control: invalid state, READY with IN request\n");
+            }
+            break;
+        default:
+            DEBUG("usbus_control: Invalid state\n");
+            assert(false);
+            break;
+    }
+    return 0;
+}
+
+/* USB endpoint 0 callback */
+static void _handler_ep0_event(usbus_t *usbus, usbus_handler_t *handler,
+                              usbus_event_usb_t event)
+{
+    usbus_control_handler_t *ep0_handler = (usbus_control_handler_t *)handler;
+
+    (void)usbus;
+    switch (event) {
+        case USBUS_EVENT_USB_RESET:
+            DEBUG("usbus_control: Reset event triggered\n");
+            ep0_handler->setup_state = USBUS_SETUPRQ_READY;
+            _usbus_config_ep0(ep0_handler);
+            break;
+        default:
+            break;
+    }
+}
+
+static void _handler_ep0_transfer(usbus_t *usbus, usbus_handler_t *handler,
+                                 usbdev_ep_t *ep, usbus_event_transfer_t event)
+{
+    usbus_control_handler_t *ep0_handler = (usbus_control_handler_t *)handler;
+
+    switch (event) {
+        case USBUS_EVENT_TRANSFER_COMPLETE:
+            _handle_tr_complete(usbus, ep0_handler, ep);
+            break;
+        default:
```

Comment:
missing default assert/debug.

Reply:
Same as above, but now for endpoint events instead of usb events